mlir/lib/Dialect/SCF/Transforms/ParallelForToNestedFors.cpp

```
@@ -0,0 +1,91 @@
+#include "llvm/Support/Debug.h"
+
+namespace mlir {
+#define GEN_PASS_DEF_SCFPARALLELFORTONESTEDFORS
+#include "mlir/Dialect/SCF/Transforms/Passes.h.inc"
+} // namespace mlir
+
+using namespace mlir;
+
+LogicalResult mlir::scf::parallelForToNestedFors(RewriterBase &rewriter,
+                                                 scf::ParallelOp parallelOp,
+                                                 scf::ForOp *result) {
+
+  if (!parallelOp.getResults().empty()) {
+    parallelOp->emitError("Currently ScfParallel to ScfFor conversion "
+                          "doesn't support ScfParallel with results.");
+    return failure();
+  }
+
+  rewriter.setInsertionPoint(parallelOp);
+
+  Location loc = parallelOp.getLoc();
+  auto lowerBounds = parallelOp.getLowerBound();
+  auto upperBounds = parallelOp.getUpperBound();
+  auto steps = parallelOp.getStep();
+
+  assert(lowerBounds.size() == upperBounds.size() &&
+         lowerBounds.size() == steps.size() &&
+         "Mismatched parallel loop bounds");
+
+  SmallVector<Value> ivs;
+  auto loopNest =
+      scf::buildLoopNest(rewriter, loc, lowerBounds, upperBounds, steps);
+
+  auto oldInductionVars = parallelOp.getInductionVars();
+  auto newInductionVars = llvm::map_to_vector(
+      loopNest.loops, [](scf::ForOp forOp) { return forOp.getInductionVar(); });
+  assert(oldInductionVars.size() == newInductionVars.size() &&
+         "Mismatched induction variables");
+  for (auto [oldIV, newIV] : llvm::zip(oldInductionVars, newInductionVars))
+    oldIV.replaceAllUsesWith(newIV);
+
+  auto *linearizedBody = loopNest.loops.back().getBody();
+  Block &parallelBody = *parallelOp.getBody();
+  for (Operation &op : llvm::make_early_inc_range(parallelBody)) {
+    // Skip the terminator of the parallelOp body.
+    if (&op == parallelBody.getTerminator())
+      continue;
+    op.moveBefore(linearizedBody->getTerminator());
+  }
+  rewriter.eraseOp(parallelOp);
+  if (result)
+    *result = loopNest.loops.front();
```

Maybe return `FailureOr<scf::LoopNest>` for this function instead of setting the pointer value to return the new loops?